src/stan/services/sample/hmc_nuts_diag_e_adapt.hpp

```diff
@@ -133,27 +133,120 @@ int hmc_nuts_diag_e_adapt(
       interrupt, logger, init_writer, sample_writer, diagnostic_writer);
 }
 
+template <class Model, typename InitContext, typename InitInvContext,
+          typename SampleWriter, typename DiagnosticWriter>
+int hmc_nuts_diag_e_adapt(
+    Model& model, const std::vector<InitContext>& init,
+    const std::vector<InitInvContext>& init_inv_metric,
+    unsigned int random_seed, unsigned int chain, double init_radius,
+    int num_warmup, int num_samples, int num_thin, bool save_warmup,
+    int refresh, double stepsize, double stepsize_jitter, int max_depth,
+    double delta, double gamma, double kappa, double t0,
+    unsigned int init_buffer, unsigned int term_buffer, unsigned int window,
+    callbacks::interrupt& interrupt, callbacks::logger& logger,
+    callbacks::writer& init_writer, std::vector<SampleWriter>& sample_writer,
+    std::vector<DiagnosticWriter>& diagnostic_writer, size_t n_chain) {
+  if (n_chain == 0) {
+    return hmc_nuts_diag_e_adapt(
+        model, init[0], init_inv_metric[0], random_seed, chain, init_radius,
+        num_warmup, num_samples, num_thin, save_warmup, refresh, stepsize,
+        stepsize_jitter, max_depth, delta, gamma, kappa, t0, init_buffer,
+        term_buffer, window, interrupt, logger, init_writer, sample_writer[0],
+        diagnostic_writer[0]);
+  } else {
+    using sample_t = stan::mcmc::adapt_diag_e_nuts<Model, boost::ecuyer1988>;
+    std::vector<boost::ecuyer1988> rngs;
+    rngs.reserve(n_chain);
+    std::vector<std::vector<double>> cont_vectors;
+    cont_vectors.reserve(n_chain);
+    std::vector<sample_t> samplers;
+    samplers.reserve(n_chain);
+    try {
+      for (int i = 0; i < n_chain; ++i) {
+        rngs.emplace_back(util::create_rng(random_seed, chain + i));
+        cont_vectors.emplace_back(util::initialize(
+            model, init[i], rngs[i], init_radius, true, logger, init_writer));
+        samplers.emplace_back(model, rngs[i]);
+        Eigen::VectorXd inv_metric = util::read_diag_inv_metric(
+            init_inv_metric[i], model.num_params_r(), logger);
+        util::validate_diag_inv_metric(inv_metric, logger);
+
+        samplers[i].set_metric(inv_metric);
+        samplers[i].set_nominal_stepsize(stepsize);
+        samplers[i].set_stepsize_jitter(stepsize_jitter);
+        samplers[i].set_max_depth(max_depth);
+
+        samplers[i].get_stepsize_adaptation().set_mu(log(10 * stepsize));
+        samplers[i].get_stepsize_adaptation().set_delta(delta);
+        samplers[i].get_stepsize_adaptation().set_gamma(gamma);
+        samplers[i].get_stepsize_adaptation().set_kappa(kappa);
+        samplers[i].get_stepsize_adaptation().set_t0(t0);
+        samplers[i].set_window_params(num_warmup, init_buffer, term_buffer,
+                                      window, logger);
+      }
+    } catch (const std::domain_error& e) {
+      return error_codes::CONFIG;
+    }
+    util::run_adaptive_sampler(samplers, model, cont_vectors, num_warmup,
```

Comment:
Similarly to the other code, I don't like how the 1 chain and N chain implementations are separate. It seems like they could be the same code and the 1 chain thing would just have 1 chain and that's fine.

Reply:
We have two right now mostly because of cmdstan. Once we add the parallel version to cmdstan we can remove 1 chain version